NextUp-comedy/api-nc
# Development

#### in this sprint:
- [x] add new controller:
   - add new user "via admin":  `POST`- `/add-user`
   - block user:  `PATCH` -  `/block-user/:userId`
   - edit user's profile: `PATCH` -  `/edit-profile`
   - get all content "paginated":  `GET` - `/contents`
   - get all users "paginated":  `GET` - `/users`
   - match user with his/her content: `PATCH` - `/match-user-content`
    - login via google: `POST` -  `/log/google` & `POST` - `/sign/google`
- [x] move all queries onto `services` folder    
- [x] use `morgan` & `winston` for logging. 
- [x] upload edited profile picture to s3
- [x] update email service with create user template and remove verify-email step
- [x] change all statics messages & values to comes from `constants` file
- [x] add extra validation layer(helper) over received schema requests.   
- [x] move all validation checks to helpers folder:
   -  checkUserExistance 
   -   comming soon ...
- [x] add a middleware to check current user role.
- [x] add `umzug`  to generate migrations and remove the old code used to generate dummy data
- [x] embed our `db-models-nc` npm package that we cloned from our neighbor [db-models-nc repo](https://github.com/NextUp-comedy/db-models-nc)
- update `package.json` scripts with: 
  - `build:db`: Joker script; use it to drop your db if exists, then build your database & seeding it with fake data. 
  - `migrate:up`: 
  - `migrate:down`
  - `migrate:reset`:
 ------- 
 ### How to test it over:
 1. To run the project"
 - `$ npm i`
 - in your `.env.development ` add your `env vars`
 - `$ npm run build:db`
 - `$ npm run dev`
------
2. after successfully running your code, you should have this dummy data generated in your db
<img src="https://user-images.githubusercontent.com/56412800/184950867-2baa37f7-da91-4561-9a78-eb46c3f8c80e.png" height=150 width=600 />
**Whereas** :

```javascript
  const user_status_id = {
        PENDING: 1, 
        APPROVED: 2,
        REJECTED: 3,
        BANNED: 4
    }
 ``` 
``` 
admin - admin@gmail.com   | admin - approved 
zaher -  zaherabuamro@gmail.com  |  "comedian - approved"
ibrahim - ibrahim.jarada@gmail.com | "comedian - rejected"
ayman - ayman@gmail.com   | "comedian - banned"
rand - rand@gmail.com     |  "comedian - pending"
```
------
3. to interact with the process front & back:
    1. go to [`front-nc repo`](https://github.com/NextUp-comedy/front-nc) `development` branch. "pull the latest version"
    2. check the current flow for now in the demo section:
------------------    
## Demo    
  https://www.loom.com/share/a8513b10cf324908bdcdc22f394011b4
  https://www.loom.com/share/28f1305941b340e4afde518c239319c6
